examples/data-frames.dx

```diff
@@ -633,17 +683,71 @@ def pivotTable (c:Label) (t:n=>{@c:a & ...f}) (agg: List {&...f} -> {&...f'}) :
 -- | true     | false | 0              | 1/4             |
 -- | true     | true  | 0              | 0               |
 
-'### TODO: `groupBy` (groupBy)
+
+'### `groupBy` (groupBy)
+Groups the rows of a table according to a specified key selector function and
+creates a result value from each group and its key.
+The rows of each group are projected by using a specified function.
+
+def mergeUniqueSortedListsWithAux {a b} [Eq a]
+  (xlist:List (a & List b))
+  (ylist:List (a & List b)) : List (a & List b) =
+    -- This function is associative, for use in a monoidal reduction.
+    -- Assumes all xs are <= all ys.
+    -- The element at the end of xs might equal the
+    -- element at the beginning of ys.  If so, this
+    -- function removes the duplicate when concatenating the lists,
+    -- and appends the two lists of bs.
+    (AsList nx xs) = xlist
+    (AsList _  ys) = ylist
+    case last xs of
+      Nothing -> ylist
+      Just (last_x, x_b_list) -> case first ys of
+        Nothing -> xlist
+        Just (first_y, y_b_list) -> case last_x == first_y of
+          False -> concat [xlist,            ylist]
+          True ->
+            concat [allExceptLast xs,
+                    (AsList 1 [(last_x, concat [x_b_list, y_b_list])]),
+                    unsafeAllExceptFirst ys]
```

Comment:
This function is duplicated way too many times. Is there a nice generic version we could write once?

Reply:
Yes, certainly.  I'll unify these the next time I work on this.